lib/pf/ipset_cache.pm

```diff
@@ -0,0 +1,135 @@
+package pf::ipset_cache;
+
+=head1 NAME
+
+pf::ipset_cache -
+
+=cut
+
+=head1 DESCRIPTION
+
+pf::ipset_cache
+
+=cut
+
+use strict;
+use warnings;
+use Moo;
+use CHI;
+use pf::log;
+use pf::util qw(pf_run);
+
+our $logger = get_logger();
+
+has ipset_binary => (is => 'ro', default => sub { 'ipset' });
+
+has cache => (is => 'ro', builder => 1, lazy => 1);
+
+has setname => (is => 'ro', required => 1);
+
+sub _build_cache {
+    my ($self) = @_;
+    return CHI->new(driver => 'RawMemory', datastore => {}, namespace => $self->setname);
+}
+
+sub add_pairs {
+    my ($self, $ip_port_pairs) = @_;
+    my @pairs = grep {!$self->is_in_cache($_) } @$ip_port_pairs;
+    if (@pairs) {
+        unless ($self->_add_pairs_to_ipset(\@pairs)) {
+            return undef;
+        }
+        $self->_add_pairs_to_cache(\@pairs);
+    }
+    return scalar @pairs;
+}
+
+sub is_in_cache {
+    my ($self, $pair) = @_;
+    unless (ref ($pair) eq 'HASH' && exists $pair->{ip} && defined $pair->{ip} && exists $pair->{port} && defined $pair->{ip} ) {
+        return undef;
+    }
+    return $self->cache->is_valid(_format_pair($pair));
+}
+
+sub _add_pairs_to_ipset {
+    my ($self, $pairs) = @_;
+    my $setname = $self->setname;
+    my $data = join("\n", (map { "add $setname " . _format_pair($_) } @$pairs), "");
+    my $pid = open(my $ipset, "| LANG=C sudo ipset -! restore 2>&1");
+    unless (defined $pid) {
+        $logger->error("Cannot start ipset ");
+        return undef;
+    }
+    $logger->trace("ipset process pid : $pid");
+    print $ipset $data;
+    close($ipset);
+    waitpid($pid, 0);
+    return 1;
+}
+
+sub _add_pairs_to_cache {
+    my ($self, $pairs) = @_;
+    my %data = map { _format_pair($_) => 1 } @$pairs;
+    $self->cache->set_multi(\%data);
+}
+
+sub _format_pair {
+    my ($pair) = @_;
+    my $port = _format_port($pair->{port});
+    return "$pair->{ip},$port";
+}
+
+sub _format_port {
+    my ($port) = @_;
+    if ($port =~ /^\d+$/) {
+        $port = "tcp:$port";
+    }
+    return $port;
+}
+
+sub populate_cache {
+    my ($self) = @_;
+    my $setname = $self->setname;
+    my @lines = pf_run("sudo ipset list $setname 2>/dev/null", accepted_exit_status => [1]);
+
+    while (my $line = shift @lines) {
+        last if $line =~ /Members:/;
+    }
+    my @pairs;
+    while (my $line = shift @lines) {
+        if ($line =~ /(\d+(?:\.\d+){3}),((?:tcp|udp):(?:\d+))/) {
```

Comment:
Could there be a trace or debug if that doesn't match

This way we could troubleshoot any malformed pairs